test/zlib/test_zlib.rb

```diff
@@ -503,6 +504,66 @@ def test_set_dictionary
       assert_raise(Zlib::StreamError) { z.set_dictionary("foo") }
       z.close
     end
+
+    def test_multithread_deflate
+      zd = Zlib::Deflate.new
+
+      s = "x" * 10000
+      (0...10).map do |x|
+        Thread.new do
+          1000.times { zd.deflate(s) }
+        end
+      end.each do |th|
+        th.join
+      end
+    ensure
+      zd&.finish
+      zd&.close
+    end
+
+    def test_multithread_inflate
+      zi = Zlib::Inflate.new
+
+      s = Zlib.deflate("x" * 10000)
+      (0...10).map do |x|
+        Thread.new do
+          1000.times { zi.inflate(s) }
```

Comment:
This is invalid usage and we should not be encouraging it.

Reply:
https://zlib.net/zlib_faq.html#faq21

> Of course, you should only operate on any given zlib or gzip stream from a single thread at a time.

Reply:
This isn't encouraging it, merely testing it doesn't crash.  I agree the spec isn't good, as you don't get the expected result when doing it.  One possible way to avoid that would be to wrap the entire call to `inflate` and `deflate` in the mutex (and possibly other methods).  I'll experiment with that.